cmd/gwi-api/api/api.go

```diff
@@ -0,0 +1,321 @@
+	// ErrAlreadyExists is to be returned when already a member.
+	AddTeamMember(ctx context.Context, teamID, volunteerEmail string) error
+	// ListTeamMembers returns the members of the given team.
+	// Password field MUST NOT be set.
+	//
+	// ErrNotFound is to be returned when the team is unknown.
+	ListTeamMembers(ctx context.Context, teamID string) ([]Volunteer, error)
+	// RemoveTeamMember removes a volunteer from the given team.
+	//
+	// ErrNotFound is to be returned when there is no such team or team member.
+	RemoveTeamMember(ctx context.Context, teamID, volunteerEmail string) error
+
+	// CountTeamMembers returns a map of team ID -> member count.
+	CountTeamMembers(ctx context.Context) (map[string]int, error)
+
+	// RunHealthcheck should return an error when the store is not healthy.
+	RunHealthcheck(ctx context.Context) error
+}
+
+// Option can be used set an API option.
+type Option func(*API)
+
+// SetTokenExpiration sets the JWT token expiration.
+func SetTokenExpiration(expiration time.Duration) Option {
+	return func(api *API) {
+		api.tokenExpiration = expiration
+	}
+}
+
+// API contains all HTTP request handlers necessary.
+type API struct {
+	// TODO: logger is currently unused. Skipping this for time reasons,
+	//       otherwise I would be logging API errors and some debug perhaps.
+	logger      *zap.Logger
+	store       Store
+	tokenSecret []byte
+
+	// Options
+	tokenExpiration time.Duration
+}
+
+// New returns a new API instance.
+func New(logger *zap.Logger, store Store, tokenSecret []byte, options ...Option) *API {
+	api := &API{
+		logger:          logger,
+		store:           store,
+		tokenSecret:     tokenSecret,
+		tokenExpiration: DefaultTokenExpiration,
+	}
+
+	for _, opt := range options {
+		opt(api)
+	}
+
+	return api
+}
+
+// PostLogin created a user session.
+// It can be used to both sign up and log in.
+//
+// Returns:
+//   * Bad Request on invalid request body
+//   * Unauthorized on account exists but password mismatch
+//   * OK on success, in which case "token" is present in the JSON response body
+func (api *API) PostLogin(c echo.Context) error {
+	// Decode the resource.
+	var volunteer Volunteer
+	if err := c.Bind(&volunteer); err != nil {
+		return err
+	}
+
+	// Validate the body.
+	switch {
+	case volunteer.Email == "":
+		return echo.NewHTTPError(http.StatusBadRequest, "email is required")
+	case volunteer.Password == "":
+		return echo.NewHTTPError(http.StatusBadRequest, "password is required")
+	}
+
+	// Call the store.
+	ctx := c.Request().Context()
+	if err := api.store.AuthenticateOrCreateVolunteerAccount(ctx, &volunteer); err != nil {
+		if err == ErrPasswordMismatch {
+			return echo.ErrUnauthorized
+		}
+		return err
+	}
+
+	// Create a JWT token and return it.
+	token, err := api.GenerateJWTToken(&volunteer)
+	if err != nil {
+		return err
+	}
+
+	return c.JSONPretty(http.StatusOK, echo.Map{
+		"token": token,
+	}, "  ")
+}
```

Comment:
Hey Ondrej, won't be better to split this into two different use cases, one for `signin` and another one for `signup`, so responsibilities are splitted?

Reply:
Yeah, I would totally do that provided there is different functionality needed for both use cases. I mean, I am always trying to weight between what is clean and what is needed not to make the codebase explode. Here I don't have a really strong opinion. A more real use case might be easier to decide actually.

Reply:
I admit that I implemented it this way here because a single DB query or store operation is enough for both. So it was pragmatic mostly. Perhaps too pragmatic. It would be more clean and understandable to split it in any real-world scenario...